### ISSUE TITLE
[Announcement] Generic Host restricts Startup constructor injection

### ISSUE THREAD

**Issue by Tratcher** [dotnet/aspnetcore#9337]
TLDR: The only types the generic Host supports for Startup constructor injection are `IHostEnvironment`, `IWebHostEnvironment`, and `IConfiguration`. Applications using WebHost are unaffected.

In 3.0 we've re-platformed the web stack onto the generic host library. You can see the change in Program.cs in the templates:

2.x:
https://github.com/aspnet/AspNetCore/blob/5cb615fcbe8559e49042e93394008077e30454c0/src/Templating/src/Microsoft.DotNet.Web.ProjectTemplates/content/EmptyWeb-CSharp/Program.cs#L20-L22
3.0:
https://github.com/aspnet/AspNetCore/blob/b1ca2c1155da3920f0df5108b9fedbe82efaa11c/src/ProjectTemplates/Web.ProjectTemplates/content/EmptyWeb-CSharp/Program.cs#L19-L24

One key behavior change here is that Host only uses one dependency injection container to build the application, as opposed to WebHost that used one for the host and one for the app. As a result the Startup constructor no longer supports custom service injection, only `IHostEnvironment`, `IWebHostEnvironment`, and `IConfiguration` can be injected. This change was made to avoid DI issues such as duplicate singleton services getting created.

Mitigations:

 Inject services into Startup.Configure:
`public void Configure(IApplicationBuilder app, IOptions<MyOptions> options)`

[We'll add more based on requests for specific scenarios.]

**Comment by springy76** [comment]
Your mitigation does not work for me: `Startup.Configure` just doesn't have the information I need, because thats why I'm passing a complex object to the constructor of the Startup class.

In my app the entire asp-net-core block itself is just a little part of the entire app. It inherits outer services by the use of nested Autofac lifetime scopes and a bunch of delegates are being collected in earlier stages which configure aspnetcore services and routes -- all of that getting passed to the Startup class.

**Comment by Tratcher** [comment]
Constructor injected services were only needed for use in Startup.ConfigureServices. Can you give specific examples of what you're doing in ConfigureServices that consumes services? It's often possible to set up delayed actions that only resolve services when your service is resolved.

**Comment by springy76** [comment]
> It's often possible to set up delayed actions that only resolve services when your service is resolved

While this helps in many cases, here it does not, because aspnetcore is just one of many (optional) services in the app:

* The app initially only processed file based jobs. Years ago Owin has been added to provide some adhoc API surface.
* There is one big Autofac DI container which also contains services from (optional) plugins
* one of the services (similar to BackgroundWorkers) pulls a class `AspNetCoreInitializer` from DI which itself pulls `ILifetimeScope` from Autofac DI
* Startup then uses `public IServiceProvider ConfigureServices(IServiceCollection services)` to add additional dynamic services, using structures from AspNetCoreInitializer instance above, and finally uses the ILifetimeScope (from AspNetCoreInitializer again) to build a new nested Autofac lifetimescope (which inherits all outer services) which then gets returned
* Startup.Configure then uses even more data from AspNetCoreInitializer to setup (plugin provided) SignalR hubs, WebDav routes, etc.

Having this written now, I wonder if `.UseStartup<Startup>();` is a requirement or if everything the `Startup` class does can also be just done against I(Web)HostBuilder?

**Comment by Tratcher** [comment]
> Having this written now, I wonder if .UseStartup<Startup>(); is a requirement or if everything the Startup class does can also be just done against I(Web)HostBuilder?

Indeed, Startup is not a required construct, it's a holdover from the OWIN days when there might not have been a Program.Main. It can be still useful for organizing code. Are you able to do everything you need to with I(Web)HostBuilder?

**Comment by springy76** [comment]
Since the AspNetCoreInitializer class I was speaking of already contains the code which calls `new WebHostBuilder()` this will be fine. Thank you for clarification.

**Comment by springy76** [comment]
Hmm. In my `Startup` class I have this method `public IServiceProvider ConfigureServices(IServiceCollection services)` which returns the DI container to use. How would I do this with IWebHostBuilder?

**Comment by davidfowl** [comment]
That does not work either. It doesn't compose with other methods on the `IHostBuilder`. There's a single DI container now that gets build after the application is fully composed.

You can now plug in a DI container using the `IServiceProviderFactory`:

```C#
public static IHostBuilder CreateHostBuilder(string[] args) =>
    Host.CreateDefaultBuilder(args)
        .ConfigureWebHostDefaults(webBuilder =>
        {
            webBuilder.UseStartup<Startup>();
        })
        .UseServiceProviderFactory(new AutofacServiceProviderFactory());
```

**Comment by NinoFloris** [comment]
And what about WebHostBuilderContext why isn't that injectable as its constituent parts are?

**Comment by Tratcher** [comment]
@NinoFloris why would you want the WebHostBuilderContext? As you say, all of its fields are available individually.

**Comment by NinoFloris** [comment]
No reason in particular, separate is fine. I don't know, it may have felt a bit inconsistent as WebHostBuilder ConfigureServices does pass it.

**Comment by grecosoft** [comment]
Using the generic host builder, how do you log?  Within the ConfigureServices method how do you write a log based on the configured logging?  How can you reference ILogger or ILoggerFactory?  Logging is important and should be easily accessible after the ConfigureLogging method is called.

**Comment by Tratcher** [comment]
@grecosoft logging is built on dependency injection, and ConfigureLogging is just a wrapper for ConfigureServices. It's not possible to log until after all of the services have been registered and the container has been built.

**Comment by grecosoft** [comment]
Then how would you write logs.  For example, if you wanted to write a log as to which service type was register?  I understand you can access after the container is created.  But how do you write logs as the application is being built?

**Comment by Tratcher** [comment]
@grecosoft it's not supported in this model, or at least not until after ConfigureSerivices. Startup.Configure is one of the earliest phases where DI services like logging are available.

**Comment by grecosoft** [comment]
Ok.  Seems to be limiting.  Just seems that logging is so important that it should be created up front and allowed to be used anywhere.  Is it possible to create a logger to be used during configuration?

**Comment by Tratcher** [comment]
No, the logger depends on configuration.

What your asking for requires a logger that is configured directly in code and doesn't use DI, Config, Options, etc.. That kind of logger is possible, but of very limited usefulness.

**Comment by grecosoft** [comment]
Tratcher - I understand.  Not an issue just need to make a small refactor to our code.  Thanks for the answering my questions and for such a fast response.

**Comment by dustinlacewell** [comment]
This is a real disappointment. By using the WebHost container, I was able to register a number of classes as "BindingServices" or "StartupServices" via attributes. Then I could use `IEnumerable<IBindingService>` and `IEnumerable<IStartupService>` in `ConfigureServices` and `Configure` respectively and my startup looked like this:

```C#
    public class Startup {

        private ILogger<Startup> _logger { get; set; }
        private IEnumerable<IStartupBinder> _binders { get; }


        public Startup(IConfiguration config, ILogger<Startup> logger, IEnumerable<IStartupBinder> binders) {
            _logger = logger;
            _binders = binders;
        }

        public void ConfigureServices(IServiceCollection services) {
            foreach (var startupBinder in _binders) {
                _logger.LogInformation($"Starting binder: {startupBinder.GetType().Name}");
                startupBinder.Bind(services);
            }
        }

        public void Configure(IApplicationBuilder app, IEnumerable<IStartupService> startupServices) {
            foreach (var startupService in startupServices) {
                _logger.LogInformation($"Starting service: {startupService.GetType().Name}");
                startupService.Startup(app);
            }
        }

    }
```

Which our team found to be a very clean design. We had a chuckle at the response that logging is simply "not supported in this model". Please reconsider whether this is the ideal way forward.

We certainly were not confused by the WebHost having a separate container. Maybe documentation was the answer instead of this backwards-incompatible change...

**Comment by wpbrown** [comment]
Just saying don't log until Configure feels wrong. I was previously injecting IOptions and ILoggers in to ConfigureServices that were known to be already fully configured during the WebHost build process. This worked well. FWIW the docs for 3.0 Startup still say you can inject a LoggerFactory.

**Comment by nblumhardt** [comment]
@wpbrown if you use Serilog, you can [initialize it before building the host](https://github.com/serilog/serilog-aspnetcore#instructions), and use it directly until it's wired into the _Microsoft.Extensions.Logging_ subsystem during start-up.

This does mean using configuration directly from the files or environment, of course, since the configuration subsystem won't have been initialized, yet; e.g. `WriteTo.File(path: Environment.GetEnvironmentVariable("MY_LOG_FILE_PATH")`. Works for me. HTH!

**Comment by francoishill** [comment]
This seems like a viable workaround.

```cs
using (var scope = services.BuildServiceProvider().CreateScope())
{
    var logger = scope.ServiceProvider.GetRequiredService<ILogger<Startup>>();
    logger.LogInformation("Application is starting up");
}
```

Important notes:

* Ensure you call `AddApplicationInsightsTelemetry` before this
* If you register any `ITelemetryInitializer`'s and want them to fire before logging, ensure to register them beforehand as well

You can also create a method to make its usage cleaner:

```cs
private static void WriteLog(IServiceCollection services, Action<ILogger> action)
{
    using (var scope = services.BuildServiceProvider().CreateScope())
    {
        var logger = scope.ServiceProvider.GetRequiredService<ILogger<Startup>>();
        action(logger);
    }
}

public void ConfigureServices(IServiceCollection services)
{
    ...
    WriteLog(services, logger => logger.LogInformation("Application starting up"));
    ...
}
```

**Comment by Tratcher** [comment]
Calling BuildServiceProvider is explicitly _NOT_ recommended, it re-creates the exact problem we were trying to solve. Building the container multiple times means you end up with multiple instances of objects that were supposed to be singletons like the logging infrastructure. It can cause numerous problems in your app. Rebuilding the container for every log statement is even worse. @francoishill please redact that post, that advice is dangerous/harmful.

**Comment by francoishill** [comment]
> Calling BuildServiceProvider is explicitly _NOT_ recommended, it re-creates the exact problem we were trying to solve. Building the container multiple times means you end up with multiple instances of objects that were supposed to be singletons like the logging infrastructure. It can cause numerous problems in your app. Rebuilding the container for every log statement is even worse. @francoishill please redact that post, that advice is dangerous/harmful.

@Tratcher I hear your argument. This advice is not to be used to continually write logs for the lifetime of the application or inside some loop. It should be used for a few startup logs. I will edit my post accordingly.

There is, however, no other solution that seems to work. Do you have any other suggestions that would work to get the Startup logs into App Insights? Other suggestions did not work with `<TargetFramework>netcoreapp2.2</TargetFramework>` and `<PackageReference Include="Microsoft.ApplicationInsights.AspNetCore" Version="2.8.0" />`. None of the logs in Startup reach App Insights.

**Comment by cheesi** [comment]
I'm also missing the logging in the startup class. Kind of disappointing... Is there really no practicable in-build solution?

**Comment by wpbrown** [comment]
I was [originally unsure](#issuecomment-514826039) what to do about this change. It forced me to rethink my approach, and I actually like this new model better. I chose to create a bare minimum "bootstrapping" ILogger ([example](https://github.com/wpbrown/oakproxy/blob/a68844d95509d5b4ea93350dcaccab1132b01532/src/OAKProxy/Program.cs#L51)), that is just used until the application is built. In my case it just logs everything to the Console or Windows Event Log and is not effected by end-user configuration (which could be considered a good thing for diagnosing application startup issues). If folks agree with this concept it may be worth suggesting in the docs.

@francoishill if you want to take the bootstrap logger concept above and also work with ApplicationInsights you can just call [AddApplicationInsights](https://github.com/microsoft/ApplicationInsights-dotnet-logging/blob/30aa32d3367f247f37582848e902e13ac605e9c6/src/ILogger/ApplicationInsightsLoggingBuilderExtensions.cs) on the ILoggingBuilder when you create your factory. App Insights normally filters to Warning level by default. If you want that on your logger too you can add the rule:
```csharp
new LoggerFilterRule("Microsoft.Extensions.Logging.ApplicationInsights.ApplicationInsightsLoggerProvider", null, LogLevel.Warning, null)));
```

**Comment by francoishill** [comment]
Very nice @wpbrown, this approach works well and feels much more "native". The only gotcha being the InstrumentationKey is not automatically obtained. This means you need to get it from the config, for example with `var instrumentationKey = _configuration.GetValue<string>("ApplicationInsights:InstrumentationKey");`. Below is my approach in Startup to achieve this.

```cs
private readonly bool _isDevelopment;
private readonly IConfiguration _configuration;
private readonly ILogger _logger;

public Startup(IConfiguration configuration, IWebHostEnvironment env)
{
    _configuration = configuration;
    _isDevelopment = env.IsDevelopment();

    _logger = GetEarlyInitializationLogger();
}

private ILogger GetEarlyInitializationLogger()
{
    var instrumentationKey = _configuration.GetValue<string>("ApplicationInsights:InstrumentationKey");
    using var loggerFactory = LoggerFactory.Create(builder =>
    {
        builder.AddConsole();
        builder.AddApplicationInsights(instrumentationKey);
    });

    return loggerFactory.CreateLogger("Initialization");
}

// This method gets called by the runtime. Use this method to add services to the container.
public void ConfigureServices(IServiceCollection services)
{
    _logger.LogInformation("Application starting up");
}
```

**Comment by alistairjevans** [comment]
Just to add to the set of workarounds; if you use NLog then it's really easy to get a normal ILogger instance in the Startup class which you can use as you did before:

```csharp
public class Startup
{
  public Startup(IConfiguration configuration)
  {
    // Get the factory for ILogger instances.
    var nlogLoggerProvider = new NLogLoggerProvider();

    // Create an ILogger.
    Logger = nlogLoggerProvider.CreateLogger(typeof(Startup).FullName);
  }
  
  public void ConfigureServices(IServiceCollection services)
  {
    Logger.LogInformation("Registering Services");
    // And the rest
  }
}
```
Some additional details are in my post about it: https://alistairevans.co.uk/2019/10/04/asp-net-core-3-0-logging-in-the-startup-class-with-nlog/

**Comment by Kaelum** [comment]
@Tratcher I'm sorry, but none of the options that you have presented, are acceptable.  Logging must be available the moment Main() is started, and should be carried through to the end of the application.  If this means that logging that needs to be handled differently then everything else, then so be it.

@alistairjevans there is a one liner that can be used for NLog anywhere within your application:

Logger = LogManager.GetCurrentClassLogger();

Until this is fixed properly in IHost, the above line is the only true work-around.  I can't believe logging during StartUp is so screwed up.

**Comment by alistairjevans** [comment]
@Kaelum, the reason I didn't use ``GetCurrentClassLogger`` is because it returns an NLogger log abstraction, whereas I wanted to get hold of the proper .NET Core ``ILogger`` type.  That way I can pass it to other objects that I might want to DI later, without leaking NLog into the rest of the app.

**Comment by Kaelum** [comment]
@alistairjevans don't forget to implement `IDisposable`, as the `NLogLoggerProvider` object needs to be disposed of.  If you later replace `Logger` with the value from the `Configure()` method, you'll need to dispose of it there as well.  That's a little too convoluted for my taste.

Unless you are going to pass the `NLogLoggerProvider` object around, I'm not sure what you mean by "I can pass it to other objects".  The logger will be tied to the object type that you created it under, unless you are saying that you're going to create additional ILogger objects for each class that you pass it to?  Passing a single `ILogger` to other objects would cause NLog rules confusion issues.

**Comment by davidfowl** [comment]
> Until this is fixed properly in IHost, the above line is the only true work-around. I can't believe logging during StartUp is so screwed up.

You can do something like this:
```C#
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;

namespace WebApplication356
{
    public class Program
    {
        public static void Main(string[] args)
        {
            using var loggerFactory = LoggerFactory.Create(builder =>
            {
                builder.AddConsole();
                builder.AddEventSourceLogger();
            });

            CreateHostBuilder(args).ConfigureServices(services => services.AddSingleton(loggerFactory)).Build().Run();
        }

        public static IHostBuilder CreateHostBuilder(string[] args) =>
            Host.CreateDefaultBuilder(args)
                .ConfigureWebHostDefaults(webBuilder =>
                {
                    webBuilder.UseStartup<Startup>();
                });
    }
}
```

There are some downsides though:
- Logger providers added via DI won't work (since you took control)
- You'll need to replicate [this](https://github.com/aspnet/Extensions/blob/359a08509cfca1901a42e97f854c3318b084e848/src/Hosting/Hosting/src/Host.cs#L93-L115) in your setup if you want to match host defaults. This gets messy since you need configuration to exist.

**Comment by Kaelum** [comment]
@davidfowl I think that the entire idea of creating the `LoggerFactory` within the `HostBuilder` is doomed for failure.  Instead, the logger should be completely defined prior to the `HostBuilder`, and then used by the `HostBuilder` as it was created.  This is why NLog and log4net allow you to configure them via a separate configuration file, and does not require you to even share the application configuration.  They do support using the application configuration file, but it isn't recommended for this specific reason.

`LoggerFactory` should support the use of standalone configuration files, including the standalone files that NLog and log4net use.  This makes significantly more sense than what we currently have.  There is no reason that I can see as to why we are forced to use the application configuration for our logging configuration.  They should be completely separate if we need them to be.

We're also moving away from JSON configuration files and back to XML configuration files, solely due to the reason that XML files transform better than JSON files do.  It also doesn't make sense to wrestle with JSON file sudo transformations, when XML transformations are so easy to work with  when using SlowCheetah.

Anywho, hopefully some day logging will stand alone.

**Comment by WhatFreshHellIsThis** [comment]
Anyone sitting on the pointy end of tech support can attest that they often need to see a **COMPLETE** startup log file from Main() line zero all the way to termination.

**Comment by davidfowl** [comment]
> Anyone sitting on the pointy end of tech support can attest that they often need to see a COMPLETE startup log file from Main() line zero all the way to termination.

You can use the code provided here to do so https://github.com/aspnet/AspNetCore/issues/9337#issuecomment-539859667

**Comment by macromania** [comment]
I have been relying on `ILogger` mainly when I started to use CosmosDb especially around registering `CosmosClient`, `CosmosContainer` which includes some steps like creating if not exists etc. Having `ILogger` in the startup for these operation was really helpful. I also get why this has been a good trade off for overall startup health of the app.

In this context, I know this is not the most elegant solution and not as structured as `ILogger` but I ended up using good old `Console.WriteLine` for critical startup logs. When I upgraded an existing .NetCore 2.2 and still wanted to continue logging these details for operational requirements, replacing `ILogger` calls with `Console.WriteLine` helped me a lot during development! 

I'm still not 100% though if these logs can be picked up and seen in Azure App Service logs but for now I hope this could help others in the meantime.

**Comment by dazinator** [comment]
My use case for injecting a custom service into `startup` class constructor, is different from the logging case. You can currently inject `IHostEnvironment` but that is not good enough for my case as I want extended information about the Host to be available when Configuring services. I want to inject my custom `IHostInfo` service which has additional information about the environment such as the configured storage directory, and other things. This extended information is needed in order to Configure Services properly. Right now my workaround is to bypass DI and use a static, but this isn't the clean feel expected from a shiny new aspnet core application!

**Comment by dazinator** [comment]
I'm not suggesting building the container twice either, I am suggesting being able to pass instances for injection via `UseStartup<>` would be enough, e.g:

```
b.UseStartup<Startup>(hostInfo);
```

**Comment by davidfowl** [comment]
That might be reasonable. To pass a separate collection or bag of values to Startup.

**Comment by springy76** [comment]
Yes please. These instances would then be available as singleton-scoped during entire lifetime?

**Comment by dazinator** [comment]
@springy76 
> These instances would then be available as singleton-scoped during entire lifetime?

I dont think this overload should automatically register these instances passed in this way for DI as singletons. You may not want them registered with DI at all. It will just use them when activating the startup type in addition to the inbuilt instances (IHostEnvironment etc) that are already supported. If you wanted these objects to also be registered as singletons with DI you can do that fairly easily by registering then yourself, in addition to passing them in this manner. I'e in ConfigureServices within your startup class or within `Host.ConfigureWebHostDefaults`

**Comment by alistairjevans** [comment]
Perhaps a simple solution here might be to allow the Startup class to be instantiated via a Lambda, so you can pass whatever you want to the constructor of your own Startup class? Something like:

```csharp
// Get whatever logger you want
var myLogger = new SomeLogger();

// Create the host
var host = Host.CreateDefaultBuilder(args)
    .UseServiceProviderFactory(new AutofacServiceProviderFactory())
    .ConfigureWebHostDefaults(webHostBuilder =>
    {
      // Configure the startup, but with a lambda
      webHostBuilder
        .UseStartup((hostingEnvironment, configuration) => new Startup(hostingEnvironment, myLogger))
        .UseContentRoot(Directory.GetCurrentDirectory());
    })
```

Arguments to the lambda could be the things available to .NET at that time, i.e. hosting environment and configuration, then you add your own things.

**Comment by Tratcher** [comment]
@alistairjevans at that point why not just use the IWebHostBuilder Configure and ConfigureServices APIs directly and capture what you need? Startup doesn't offer much beyond those.

**Comment by dazinator** [comment]
Some extended info about how I'd plan to use this feature (if `UseStartup<TStartup>° had an overload that took additional object instances to be used when activating the type) and also it covers why a lambda wouldnt be helpful to me)

During web app startup I commonly find I need to repeat certain tasks. For example, I like to configure a location where files can be persisted accross deployments. I dont want this to be inside the typical ContentRoot or WebRoot location because if I deploy a new version of the app alongside the old version, suddenly all those files I may have persisted in the old app's content root are gone. Secondly sometimes web apps are run as a farm. All instances of the web app need access to persisted these files so therefore (unbeknownst to application code) I often set this persistent content directory up as file share like an azure file share etc.  Secondly I need to configure DataProtection. The easiest option i(least dependencies) is to persist the keys to this configured persistent storage location. That way the keys aren't lost when I deploy a new version of the app, and if it's a file share that all web nodes can see then it works in the farm scenario too. By when configuring DataProtection to persist keys it liks to use the AppData folder by default but this is no good if you then want to change to a web farm scenario - I like to set things up in a way that will mean no code changes should I want my app to  run as a farm. Thirdly I like to configure JWT bearer authentication. To do this I need to generate and then persist another key, used to sign the JWT token. You guessed it, I need to save this key to the same persistent content location. I encrypt this key using DataProtection. Therefore there is a chain of trust with the DataProtection key at the root. Now.. I realise this is getting rather long, so I'll stop here but you can see that there a bunch of concerns I commonly need to take care off before I can think about having an app that is flexible enough to work solo, or in a web farm with no code changes necessary. So with this in mind..

I'd ship some nuget package that has an extension method for WebHostBuilder in it. Something like `UseOpinionatedStartup<TStartup>()`

Internally, that method would call a bunch of methods on WebHostBuilder like ConfigureServices etc to add my opinionated default services and set up plumbing level and boiler plate stuff so that I dont have to repeat it in all my apps startup classes. Likewise if you already have a startup class you can use my nuget package and call this extension method, without changing your startup class. Just know you now no longer need to worry about setting up certain  things like data protection inside your startup class because you've done it in my opinionated way.

Now I can do all of that software far today. The thing I cant currently do, is, as a consumer of this nuget package, when calling `webhostBuilder.UseOpionatedStartup<MyStartup>()` I'd like to get access to some information about this opinionated startup such as the persistent content directory and the like, as I want to configure some additional services to use that location. However even though that extension method may register particular services (like IHostInfo) these cant be injected into my startup class so I cant get access to them yet. Likewise the Extension method calls `UseStartup<TStartup>` internally, but it has no way to pass this information to that startup class. The suggestion I made would allow my extension method to do something like this:

```
var hostInfo = ConfigureHostOpiniinated();
webHostBuilder.UseStartup<TStartup>(hostInfo);


```

Now it's up to the consumer if they need access to this opinionated startup information then they can add an argument to their constructor to have my HostInfo injected. If they dont need it, they dont need that constructor argument and it will not error when activating their startup class (The HostInfo I pass to activate the startup class would just be ignored if not needed for the constructor)

You can see also why a lambda doesnt help my case - at the point my Extension method activates the startup class it knows nothing of its constructor. I could build my own activation logic ofcourse, but I'd prefer if this useful feature was inbuilt.

**Comment by MV10** [comment]
@dazinator If I'm following you, it seems you could simply skip the use of `Startup` (which is just a convention, as mentioned earlier) and store the various pieces of shared information in the `IHostBuilder.Properties` collection under one or more pre-defined keys in your `UseOpinionatedStartup` implementations. That's pretty much what the `Properties` collection is for -- it isn't passed to `Startup` but it's basically the "bag of values" that davidfowl mentioned.

As for the logging issue, I'm starting to hear from people who are realizing there is a problem. We have a lot of headless jobs that rely solely on logging for progress and failure reporting. I've seen a few cobbled-together work-arounds. People need this so they _will_ try to find an answer, which may be worse than the problem that was meant to be solved. Building early and breaking DI is not a solution. I'm trying to devise an in-house standard along the lines of caching messages until the builder runs. Exceptions actually run a new logger-only builder (which hopefully doesn't fail itself) merely to dump the cache and the exception ... but it feels like a hack, and of course, this lightweight pseudo-logger lacks the many important features of a real modern logger.

+1 for the observations that logging is a special-case / cross-cutting and if it needs special handling, so be it. This is non-trivial.

**Comment by MV10** [comment]
I spent the day cleaning up my implementation of reliable basic logging during Host Builder configuration. I still need to get it packaged for public use, but maybe somebody will find it useful.

https://github.com/MV10/GenericHostBuilderLogger

**Comment by MichaelPuckett2** [comment]
How do you pass command line args to Startup?

**Comment by Tratcher** [comment]
@MichaelPuckett2 inject IConfiguration, that will contain your command line args if you provided them to the host builder.

**Comment by razzemans** [comment]
I also run into issues with this. We are setting a custom `ISecureDataFormat` in the `CookieAuthenticationOptions` and running into a production issue where decoding the data fails at random - I really like to pass a logger to our implementation of the `ISecureDataFormat` to see what is going wrong. In my opinion this should be trivial and I am disappointed that I don't have a way of doing that at the moment.

```
public void ConfigureServices(IServiceCollection services)
{
	services.AddAuthentication(o =>
	{
		o.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
	})
	.AddCookie(o =>
	{
		o.TicketDataFormat = new CustomDataFormat(settings.Secret, /* Add logging? */);
	});
}
```

**Comment by Tratcher** [comment]
@razzemans does this help?
```
            services.AddAuthentication(options =>
            {
                options.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
            }).AddCookie(CookieAuthenticationDefaults.AuthenticationScheme);

            services.AddTransient<IConfigureOptions<CookieAuthenticationOptions>>(sp =>
                new ConfigureNamedOptions<CookieAuthenticationOptions>(CookieAuthenticationDefaults.AuthenticationScheme, o =>
                {
                    o.TicketDataFormat = new CustomDataFormat("secret", sp.GetRequiredService<ILogger<CustomDataFormat>>());
                }));
```
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-3.1#use-di-services-to-configure-options
https://github.com/dotnet/aspnetcore/blob/6255c1ed960f5277d2e96ac2d0968c2c7e844ce2/src/Security/Authentication/Cookies/src/CookieExtensions.cs#L29
https://github.com/dotnet/extensions/blob/cc317e6112479d36ba77a57e1eeea5d88a164a6c/src/Options/Options/src/OptionsServiceCollectionExtensions.cs#L213
https://github.com/dotnet/extensions/blob/cc317e6112479d36ba77a57e1eeea5d88a164a6c/src/Options/Options/src/OptionsBuilder.cs#L73-L74

We're missing some overloads of AddCookie or Configure that would make this scenario easier.
```
.AddCookie(CookieAuthenticationDefaults.AuthenticationScheme, (o, serviceProvider) =>
{
        o.TicketDataFormat = new CustomDataFormat("secret", sp.GetRequiredService<ILogger<CustomDataFormat>>());
});

.AddCookie<ILogger<CustomDataFormat>>(CookieAuthenticationDefaults.AuthenticationScheme, (o, logger) =>
{
        o.TicketDataFormat = new CustomDataFormat("secret", logger);
});

services.Configure<CookieAuthenticationOptions, ILogger<CustomDataFormat>>(CookieAuthenticationDefaults.AuthenticationScheme, o =>
{
       o.TicketDataFormat = new CustomDataFormat("secret", sp.GetRequiredService<ILogger<CustomDataFormat>>());
});
```

Let me know which of these look useful and I'll file an issue.

**Comment by razzemans** [comment]
Thanks a lot @Tratcher - that indeed works.

I think I would prefer your first proposal - it is also consistent with the factory methods overloads of the `IServiceCollection` interface.

**Comment by IanKemp** [comment]
I'm with the other people in this thread calling for a special-case exemption for logging: it's absolutely necessary. The fact that this apparently wasn't considered at all during Core 3.0 dev is absolutely bizarre, in my opinion.

I understand the rationale behind not doing it, and the hoops that @MV10 jumped through to make a usable workaround that doesn't break everything shows how difficult it would be to do without the "fake" `HostBuilder` from 2.x, but this really feels like a case of the programmers being forced to serve the code, not the opposite.

Please strongly consider adding this capability back as soon as possible. Fobbing people off with half-a**ed workarounds, or requiring them to install third-party NuGet packages to get this basic functionality back, **is not an acceptable solution in any way shape or form**. Usability needs to trump ideology here.

**Comment by dazinator** [comment]
To echo @IanKemp, and add some info, there is nothing stopping us right now from initialising logging early (creating an ILoggerProvider instance, be that console, serilog etc) immediately in program main, and using that to create an `ILogger` instance and log with right away in Program.cs. This is great to catch and log host startup errors- just remember to dispose of ILoggerProvider in a finally block so it has a chance to flush any messages before the app terminates. This approach works in conjunction with AddLogging() which happens sometime later - you can call AddProvider() to add your instance of your ILoggerProvider that you initialised early in Program main so that the same instance is available for DI.

All good so far. A few issues remain for the framework to address though in my view:

1. It should actively encourage your to initialise a logger as the first thing you do. If the framework knows about this initial ILoggerProvider instance (which it should imho) then it could also automatically include it in the Providers list where AddLogging is called (it already includes other loggers by default) and it could use this logger itself when doing any host level startup stuff.

2. When using startup classes, the host doesnt allow you to activate them passing additional constructor arguments. If it did then we could optionally supply an ILogger instance that we initialised early in program.cs, and our logic in startup.cs could now have the benefit of being able to log. I work around this currently by referencing the ILogger from a static field e.g Program.Logger but this hurts testability of startup.cs.

To address that part we'd just need the ability to supply additional constructor parameter values for activating startup classes, so we could do:

```
HostBuilder.UseStartup<TStartup>(loggerInstance)
```

**Comment by davidfowl** [comment]
> It should actively encourage your to initialise a logger as the first thing you do. 

Sounds like you're suggesting the templates change to do this. 

> If the framework knows about this initial ILoggerProvider instance (which it should imho) then it could also automatically include it in the Providers list where AddLogging is called (it already includes other loggers by default) and it could use this logger itself when doing any host level startup stuff, including allowing this ILogger to be injected into any startup class.

This requires some more thought but it might work if we do something hacky like try to detect an existing ILoggerProvider. There are still 2 DI containers in this universe so for example, you'd need to make sure to configure options for logging in the right place as the streams don't cross.

> When using startup classes, the host doesnt allow you to activate them supplying your own additional constructor parameters over and above what it knows how to inject natively. If it did then we could optionally supply an ILogger instance or anything else that we initialised early in program.cs, and our logic in startup.cs could now have the benefit of being able to log. I work around this currently by referencing the ILogger from a static field e.g Program.Logger but this hurts testability of startup.cs.

This sounds reasonable and it's something we do today for middleware (see UseMiddleware). 

@danzinator Can you file 2 separate issues for these items? No promises of course but having 2 concrete proposals would help the discussion.

**Comment by dazinator** [comment]
@davidfowl ok, have created #19807 and #19809
